### ISSUE TITLE
proposal: Implement a feature rich CLI parser for Go

### ISSUE THREAD

**Issue by satyenr** [golang/go#12893]
While the current flag library is good enough for many purposes, I sometimes find it too barebones. Here are the things I am looking for in particular:

* Support for GNU style `--long=LONG` options.
* Ability to define the prefix character which allows implementing Windows styple /f switches.
* Support for mixing positional arguments with options. In fact, unless I am much mistaken, there is no support for parsing and validating positional arguments at all.
* A way to make a flag mandetory or optional.
* A way to restrict the choices a flag can have.
* A way for a flag to have multiple values.
* A way to read composite values and automatically map the components to the fields of a struct.

To summarise, I am looking for an argument parser on par with or more feature rich than Python's argparse. While there are some third party libraries that provide these features (or I could roll out my own), I feel that a powerful CLI parser should be a part of the standard library.

Whether it is implemented by extending the flags package or as a completely new library is secondary.

**Comment by satyenr** [comment]
Unless there are any strong objections to bundling such a library with Go, or someone is already working on such a library, I'd like to write a more detailed description of the propsed library and possibly implement it.

**Comment by bradfitz** [comment]
This has been proposed and rejected in the past. The Go flags package intentionally chooses not to implement lots of features. Many alternatives exist; see https://godoc.org/?q=flag for some of them.

Related: https://golang.org/doc/faq#x_in_std

/cc @adg since it's a proposal, but I think it's safe to close this.
